### PR TITLE
Add create project page

### LINKED ISSUE #71: New "Create Project Page" to select instruments and store them in system card
A new page in the TAD/AMT.

Note that this page may be temporary (and eventually replaced by a landing page, directly filled from the shopping card in the AK).

Image

The instruments are loaded on the backend from the Instrument Registry (we don't need to check schema yet).

On "Create Project" a System Card with the selected instruments (their URNs and versions) is created in a new field "selected_instruments". The project name is also written to the System Card.

Other fields in the system card can be left empty (for now).

We make an entry in the database that links to the system card and has no other information.

After creation, the user lands on the task page with the project selected with the name visible (and thus loaded from the system card). It is ok (for now) to just load the latest created project.

The functionality is covered in a backend test.

### PR DESCRIPTION
**Pull request by berrydenhartog**
# Description

Add project page with instruments

Resolves #71 

## Checklist

Please check all the boxes that apply to this pull request using "x":

- [x]  I have tested the changes locally and verified that they work as expected.
- [x]  I have followed the project's coding conventions and style guidelines.
- [x]  I have rebased my branch onto the latest commit of the main branch.
- [x]  I have squashed or reorganized my commits into logical units.
- [x]  I have read, understood and agree to the [Developer Certificate of Origin](../blob/main/DCO.md), which this project utilizes.

### CODE REVIEW

File: tad/api/routes/projects.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    storage_writer = WriterFactory.get_writer(
+        writer_type="file", location=system_card_file.parent, filename=system_card_file.name
+    )
+    storage_writer.write(system_card.model_dump())
+
+    # TODO(Berry): Parse instruments and add them to the project
+
+    headers = {"HX-Location": f"/project/{project.id}"}
+    return templates.TemplateResponse(request, "redirect.html.j2", headers=headers)
```
Comment (anneschuth) [inline tad/api/routes/projects.py:63]: We could abstract this maybe into a "redirect response", with an interface like 
```python
templates.Redirect(request, f"/project/{project.id}")
```

Comment (berrydenhartog) [reply tad/api/routes/projects.py:63]: Nice idea

File: tad/schema/github.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from pydantic import BaseModel, HttpUrl, RootModel
```
Comment (laurensWe) [inline tad/schema/github.py:1]: Should we enforce here that only https is allowed? https://docs.pydantic.dev/latest/api/networks/#pydantic.networks.AnyHttpUrl

Comment (berrydenhartog) [reply tad/schema/github.py:1]: We have no control over it since github creates these urls

File: tests/site/static/templates/test_template_redirect.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from tad.api.deps import templates
+from tests.constants import default_fastapi_request
+
+
+def test_tempate_redirect():
```
Comment (anneschuth) [inline tests/site/static/templates/test_template_redirect.py:5]:
```suggestion
def test_template_redirect():
```

File: tests/site/static/templates/test_template_new_project.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+from tad.api.deps import templates
+from tests.constants import default_fastapi_request
+
+
+def test_tempate_projects_new():
```
Comment (anneschuth) [inline tests/site/static/templates/test_template_new_project.py:5]:
```suggestion
def test_template_projects_new():
```

File: tests/site/static/templates/test_template_redirect.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    # given
+    request = default_fastapi_request()
+
+    # when
+    response = templates.TemplateResponse(request, "redirect.html.j2")
+
+    # then
+    assert response.headers["Content-Type"] == "text/html; charset=utf-8"
+    assert response.body == b""
```
Comment (anneschuth) [inline tests/site/static/templates/test_template_redirect.py:14]: should we test for the header to be present?

File: tests/clients/test_client_github.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import pytest
+from httpx import HTTPStatusError
+from pytest_httpx import HTTPXMock
```
Comment (anneschuth) [inline tests/clients/test_client_github.py:3]: I thought we said we'd use https://vcrpy.readthedocs.io/en/latest, see https://minbzk.github.io/ai-validation/projects/tad/adrs/0004-software-stack/

File: tests/e2e/test_create_project.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+from playwright.sync_api import Page, expect
+
+
+def test_e2e_create_project(page: Page):
```
Comment (anneschuth) [inline tests/e2e/test_create_project.py:4]:
```suggestion
def test_create_project(page: Page):
```

Or otherwise maybe prefix all e2e tests with `e2e`?

File: tad/utils/path.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+def is_safe_path(basedir: Path, path: Path, follow_symlinks: bool = True) -> bool:
+    # Resolve the absolute path of the base directory
+    base = basedir.resolve()
+
+    # Resolve the absolute path of the provided path
+    target = path.resolve() if follow_symlinks else path
+
+    # Check if the target path starts with the base directory path
+    return base in target.parents or base == target
```
Comment (anneschuth) [inline tad/utils/path.py:12]:
```suggestion
    return target.parents.startswith(base) or base == target
```

File: tests/services/test_instruments_service.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        headers={"X-RateLimit-Remaining": "7", "X-RateLimit-Reset": "200000000", "Content-Type": "application/json"},
+    )
+
+    httpx_mock.add_response(
+        url="https://raw.githubusercontent.com/MinBZK/instrument-registry/main/instruments/iama.yaml",
+        content=b"test: 1",
+        headers={"X-RateLimit-Remaining": "7", "X-RateLimit-Reset": "200000000", "content-type": "text/plain"},
+    )
+
+    # when
```
Comment (anneschuth) [inline tests/services/test_instruments_service.py:47]:
```suggestion
    # then
```

File: tests/repositories/test_projects.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    project.id = 1
+    project_duplicate = default_project()
+    project_duplicate.id = 1
+
+    project_repository.save(project)
+
+    with pytest.raises(RepositoryError):
+        project_repository.save(project_duplicate)
+
+    project_repository.delete(project)  # cleanup
```
Comment (anneschuth) [inline tests/repositories/test_projects.py:59]: I'd be slightly worried if we need this. We create a new db for each test, right?

File: tad/api/routes/projects.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    return templates.TemplateResponse(request, "projects/new.html.j2", {"instruments": instruments})
+
+
+@router.post("/new", response_class=HTMLResponse)
+async def post_new(
+    request: Request,
+    project_new: ProjectNew,
+    projects_service: Annotated[ProjectsService, Depends(ProjectsService)],
+) -> HTMLResponse:
+    system_card_file = get_settings().CARD_DIR / f"{project_new.name}_system.yaml"
```
Comment (anneschuth) [inline tad/api/routes/projects.py:40]: Name's aren't unique, are they? Wouldn't it be better to use the id?

File: compose.yml [hunk 12]
```diff
@@ -14,7 +14,7 @@ services:
     environment:
       - ENVIRONMENT=demo
     ports:
-      - 8000:8000
+      - 8070:8000
```
[the remark this answers is not in the document]
Comment (berrydenhartog) [reply compose.yml:17]: Allot of tools use 8000. wanted a different port so i can run things parallel:
vscode uses 8000
pinniped used 8000

File: tad/core/config.py [hunk 13]
```diff
@@ -34,6 +35,8 @@ class Settings(BaseSettings):
     DEBUG: bool = False
     AUTO_CREATE_SCHEMA: bool = False
 
+    CARD_DIR: Path = Path("/tmp/")  # TODO(berry): create better location for model cards  # noqa: S108
```
[the remark this answers is not in the document]
Comment (berrydenhartog) [reply tad/core/config.py:38]: Not yet

File: tests/repositories/test_projects.py (hunk 10 above)
Comment (berrydenhartog) [reply tests/repositories/test_projects.py:59]: No, we do not

File: tad/api/routes/projects.py (hunk 11 above)
Comment (berrydenhartog) [reply tad/api/routes/projects.py:40]: agreed. will change it

Comment (berrydenhartog) [reply tad/api/routes/projects.py:40]: @anneschuth. What do you think about an id and a name? think this would help if we start writing it to remove locations for usability

File: tad/clients/github.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
+import logging
+from datetime import datetime, timezone
+
+import httpx
+from tad.schema.github import RepositoryContent
+
+logger = logging.getLogger(__name__)
+
+
+class GitHubClient:
```
[the remark this answers is not in the document]
Comment (berrydenhartog) [reply tad/clients/github.py:10]: Created task: https://github.com/orgs/MinBZK/projects/7/views/1?pane=issue&itemId=71202148

File: tests/site/static/templates/test_template_redirect.py (hunk 3 above)
Comment (berrydenhartog) [reply tests/site/static/templates/test_template_redirect.py:5]: Solved

File: tests/e2e/test_create_project.py (hunk 7 above)
Comment (berrydenhartog) [reply tests/e2e/test_create_project.py:4]: Renamed all with e2e

File: tad/utils/path.py (hunk 8 above)
Comment (berrydenhartog) [reply tad/utils/path.py:12]: applied change

File: tad/api/routes/projects.py (hunk 11 above)
Comment (berrydenhartog) [reply tad/api/routes/projects.py:40]: The issue is that we did not have the ID of the project because the project is not inserted into the database. I will need to change the model that we can insert a project without a system card. will apply that now.

File: tests/repositories/test_projects.py (hunk 10 above)
Comment (anneschuth) [reply tests/repositories/test_projects.py:59]: Right. We discussed this this morning in #90

**Review by anneschuth — DISMISSED** [review]
Nice! 🚢

**Review by anneschuth — DISMISSED** [review]
:shipit:

File: tad/api/routes/projects.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    return templates.TemplateResponse(request, "projects/new.html.j2", {"instruments": instruments})
+
+
+@router.post("/new", response_class=HTMLResponse)
+async def post_new(
+    request: Request,
+    project_new: ProjectNew,
+    projects_service: Annotated[ProjectsService, Depends(ProjectsService)],
+) -> HTMLResponse:
+    project = projects_service.create(project_new.name)
```
Comment (uittenbroekrobbert) [inline tad/api/routes/projects.py:38]: nit-picking: I think (all) the steps below could be part of the project_service create method, as that would be a more logic place for 'business logic' steps than in the api call.

File: tad/api/routes/tasks.py [hunk 16]
```diff
@@ -21,21 +21,15 @@ async def move_task(
...
-            moved_task.previous_sibling_id,
-            moved_task.next_sibling_id,
-        )
-        # todo(Robbert) add error handling for input error or task error handling
-        return templates.TemplateResponse(request=request, name="task.html.jinja", context={"task": task})
-    except Exception:
-        return templates.TemplateResponse(
-            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, request=request, name="error.html.jinja"
-        )
+    # because htmx form always sends a value and siblings are optional, we use -1 for None and convert it here
```
Comment (uittenbroekrobbert) [inline tad/api/routes/tasks.py:24]: I assume removing the try/catch is possible because this exception is now handled by the 'get_undefined_behaviour' method in the LocaleJinja2Templates?

File: tad/models/project.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
+from pathlib import Path
+from typing import TypeVar
+
+from pydantic import field_validator
+from sqlmodel import Field, SQLModel  # # pyright: ignore [reportUnknownVariableType]
```
Comment (ChristopherSpelt) [inline tad/models/project.py:5]:
```suggestion
from sqlmodel import Field, SQLModel  # pyright: ignore [reportUnknownVariableType]
```

**Review by ChristopherSpelt — COMMENTED** [review]
Looks good!

File: tad/models/project.py [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+class Project(SQLModel, table=True):
+    id: int | None = Field(default=None, primary_key=True)
+    name: str = Field(max_length=255, min_items=3)
+    model_card: str | None = Field(description="Model card storage location", default=None)
+
+    @field_validator("model_card")
+    @classmethod
+    def validate_model_card(cls: type[T], model_card: str) -> str:
+        if not Path(model_card).is_dir() and not Path(model_card).is_file():
```
Comment (uittenbroekrobbert) [inline tad/models/project.py:18]: What is the use case for model_card as directory?

File: tad/schema/system_card.py [hunk 19]
```diff
@@ -0,0 +1,12 @@
+from pydantic import (
+    BaseModel,
+    Field,  # pyright: ignore [reportUnknownMemberType]
+)
+
+from tad.schema.instrument import InstrumentBase
+
+
+class SystemCard(BaseModel):
+    schema_version: str = Field(default="0.1a6")
```
Comment (uittenbroekrobbert) [inline tad/schema/system_card.py:10]: what is this default based on and is it wise to have a default at all?

File: tad/services/tasks.py [hunk 20]
```diff
@@ -49,7 +49,7 @@ def move_task(
         task = self.repository.find_by_id(task_id)
 
         if status.name == "done":
-            self.system_card.title = task.title
+            self.system_card.name = task.title
```
Comment (uittenbroekrobbert) [inline tad/services/tasks.py:52]: Maybe out of scope, but the TaskService now has a (default / demo) SystemCard based on demo properties. It would be nice if the system card for the selected project would be updated here instead.

Comment (uittenbroekrobbert) [reply tad/services/tasks.py:52]: Then again.. the (demo) tasks we have now are not relatable to the the instruments we can select, so maybe this is too unrelated.

File: tad/utils/path.py [hunk 21]
```diff
@@ -0,0 +1,12 @@
+from pathlib import Path
+
+
+def is_safe_path(basedir: Path, path: Path, follow_symlinks: bool = True) -> bool:
```
Comment (uittenbroekrobbert) [inline tad/utils/path.py:4]: Nitty picking: this is nice, but I don't see it being used except for test coverage.

File: tests/api/routes/test_projects.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+from tad.schema.project import ProjectNew
+from tad.schema.system_card import SystemCard
+from tad.services.instruments import InstrumentsService
+from tad.services.storage import FileSystemWriteService
+
+from tests.constants import default_instrument
+
+
+@pytest.fixture()
+def init_instruments() -> Generator[None, None, None]:  # noqa: PT004
```
Comment (uittenbroekrobbert) [inline tests/api/routes/test_projects.py:15]: This technique is very nice! Where possible, we should apply this more when using Mocks for "temporary original replacements".

File: tests/e2e/test_move_task.py [hunk 23]
```diff
@@ -14,53 +14,63 @@ def test_move_task_to_column(browser: Page, db: DatabaseTestUtils) -> None:
...
     """
-    Test moving a task in the browser below another task and verify that after a reload
+    Test moving a task in the page below another task and verify that after a reload
     it is in the right position in the column.
     :return: None
     """
     all_status = [*all_statusses()]
     db.given([*all_status])
 
-    task1 = default_task(title="Task 1", status_id=all_status[0].id, sort_order=10)
```
Comment (uittenbroekrobbert) [inline tests/e2e/test_move_task.py:43]: Why did you remove the sort_order? This is needed to make sure the tasks are ordered correctly, the default_task gives them all a sort_order of 1. So I am surprised this test still works?!

Comment (uittenbroekrobbert) [reply tests/e2e/test_move_task.py:43]: Ah. It probably works because you move card 1 below card 3. It won't work if you put card1 between card2 and card3.

**Review by uittenbroekrobbert — DISMISSED** [review]
Made some minor remarks, other than that, LGTM!

File: tad/api/routes/tasks.py (hunk 16 above)
Comment (berrydenhartog) [reply tad/api/routes/tasks.py:24]: No, i added better exception handling reponse

File: tad/models/project.py (hunk 18 above)
Comment (berrydenhartog) [reply tad/models/project.py:18]: Nothing

File: tad/schema/system_card.py (hunk 19 above)
Comment (berrydenhartog) [reply tad/schema/system_card.py:10]: Default is based on what is currenlty in the repo. I needed to add something because the field is required and i did not want to ask the user

File: tad/utils/path.py (hunk 21 above)
Comment (berrydenhartog) [reply tad/utils/path.py:4]: It was used, but got removed because we dont use the name anymore in the system file.

File: tad/api/routes/projects.py (hunk 15 above)
Comment (berrydenhartog) [reply tad/api/routes/projects.py:38]: I knew you where gona comment on this one ;)

Comment (berrydenhartog) [reply tad/api/routes/projects.py:38]: Moved into service

File: tad/services/projects.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+
+        self.repository.save(project)
+
+        system_card_file = get_settings().CARD_DIR / f"{project.id}_system.yaml"
+
+        project.model_card = str(system_card_file)
+
+        project = self.update(project)
+
+        instruments: list[InstrumentBase] = [
```
Comment (uittenbroekrobbert) [inline tad/services/projects.py:41]: (re-commenting because I think it got lost due to refactoring this code):
When I select 1 instrument, it loops over the string instead of the list, and thus creates a system card containing instruments of 1 letter (being all the letters of the chosen instrument urn) instead of 1 instrument with the full urn.

Comment (berrydenhartog) [reply tad/services/projects.py:41]: I will have a look. indeed missed this one

Comment (berrydenhartog) [reply tad/services/projects.py:41]: @uittenbroekrobbert . Nice find. solved it in the schema validator

**Review by anneschuth — APPROVED** [review]
LGTM!

**Review by uittenbroekrobbert — APPROVED** [review]

File: tests/site/static/templates/test_template_redirect.py (hunk 5 above)
Comment (berrydenhartog) [reply tests/site/static/templates/test_template_redirect.py:14]: No, its a template
